### ISSUE TITLE
ngcc conflicts during parallel build

### ISSUE THREAD

**Issue by junekpavel** [nrwl/nx#4937]
## Current Behavior
During parallel build, the ngcc conflicts are occurring.

## Expected Behavior
There should be no ngcc conflicts.

## Steps to Reproduce
The issue can be reproduced in nx-incremental-large-repo after updating to version >= 11.2 when running parallel build:

```
nx run-many --all --target build --parallel --maxParallel 10
```

https://github.com/junekpavel/nx-incremental-large-repo/tree/build-issues#ngcc-conflict

### Failure Logs
```
(...)

> nx run app0-lib4-childlib8:build [retrieved from cache]
Building Angular Package

------------------------------------------------------------------------------
Building entry point '@largerepo/app0/lib4/childlib8'
------------------------------------------------------------------------------
Bundling to FESM2015

------------------------------------------------------------------------------
Built Angular Package
 - from: /(...)/nx-incremental-large-repo/libs/app0/lib4/childlib8
 - to:   /(...)/nx-incremental-large-repo/dist/libs/app0/lib4/childlib8
------------------------------------------------------------------------------

> nx run app0-lib4-childlib9:build [retrieved from cache]
Building Angular Package

------------------------------------------------------------------------------
Building entry point '@largerepo/app0/lib4/childlib9'
------------------------------------------------------------------------------
Bundling to FESM2015

------------------------------------------------------------------------------
Built Angular Package
 - from: /(...)/nx-incremental-large-repo/libs/app0/lib4/childlib9
 - to:   /(...)/nx-incremental-large-repo/dist/libs/app0/lib4/childlib9
------------------------------------------------------------------------------
Another process, with id 36306, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36306, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36306, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36306, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36306, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36306, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36306, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36306, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36310, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36310, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36310, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36310, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Browserslist: caniuse-lite is outdated. Please run the following command: `yarn upgrade`
Another process, with id 36310, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36310, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36310, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36313, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36313, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36313, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36313, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36313, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36313, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36309, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Browserslist: caniuse-lite is outdated. Please run the following command: `yarn upgrade`
Another process, with id 36309, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36309, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36309, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36309, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36311, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36311, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Browserslist: caniuse-lite is outdated. Please run the following command: `yarn upgrade`

> nx run app0-lib3-childlib8:build 
Building Angular Package

------------------------------------------------------------------------------
Building entry point '@largerepo/app0/lib3/childlib8'
------------------------------------------------------------------------------
Bundling to FESM2015

------------------------------------------------------------------------------
Built Angular Package
 - from: /(...)/nx-incremental-large-repo/libs/app0/lib3/childlib8
 - to:   /(...)/nx-incremental-large-repo/dist/libs/app0/lib3/childlib8
------------------------------------------------------------------------------
Another process, with id 36311, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36311, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36312, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36312, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Another process, with id 36312, is currently running ngcc.
Waiting up to 250s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /(...)/nx-incremental-large-repo/node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Browserslist: caniuse-lite is outdated. Please run the following command: `yarn upgrade`

(...)

```

### Environment
```

  Node : 12.16.2
  OS   : darwin x64
  yarn : 1.22.0
  
  nx : Not Found
  @nrwl/angular : 11.4.0
  @nrwl/cli : 11.4.0
  @nrwl/cypress : 11.4.0
  @nrwl/devkit : 11.4.0
  @nrwl/eslint-plugin-nx : Not Found
  @nrwl/express : Not Found
  @nrwl/jest : 11.4.0
  @nrwl/linter : 11.4.0
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : Not Found
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 11.4.0
  @nrwl/web : 11.4.0
  @nrwl/workspace : 11.4.0
  typescript : 4.0.3
```

**Comment by the-ult** [comment]
Same here when using `--parallel --maxParallel 3`

```
 Node : 14.15.4
  OS   : darwin x64
  npm  : 6.14.11
  
  nx : Not Found
  @nrwl/angular : 11.4.0
  @nrwl/cli : 11.4.0
  @nrwl/cypress : 11.4.0
  @nrwl/devkit : 11.4.0
  @nrwl/eslint-plugin-nx : Not Found
  @nrwl/express : 11.4.0
  @nrwl/jest : 11.4.0
  @nrwl/linter : 11.4.0
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : 11.4.0
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 11.4.0
  @nrwl/web : Not Found
  @nrwl/workspace : 11.4.0
  typescript : 4.1.5
```

**Comment by juristr** [comment]
Yep, weirdly that started happening in Angular v11 🤔. We're going to look into it, but it shouldn't have any impact at the final outcome at least.

**Comment by the-ult** [comment]
Our Gitlab CI was choking on the ngcc errors. => Build failed 🤔

**Comment by juristr** [comment]
🤔 hmm...

meanwhile you could try to add a `ngcc.config.js` at the root of the nx workspace and add something like

```
module.exports = {
  locking: {
    retryAttempts: 50,
    retryDelay: 50,
  },
};
```

And play with the retryDelay. Maybe that helps

**Comment by the-ult** [comment]
Thanks @juristr 

Running `npx nx affected --target=build --base=origin/release --prod --parallel --maxParallel=3 --with-deps`
With the suggested `ngcc.config.js` sometimes results in:

```
//  Both on the CI and local
———————————————————————————————————————————————
>  NX   SUCCESS  Running target "codegen" succeeded
$ npx nx affected --target=build --base=origin/release --prod --parallel --maxParallel=3 --with-deps
>  NX  Running target build for projects:
  - ddb-graphql-server
  - gupo
  - uitkering
———————————————————————————————————————————————
Another process, with id 740, is currently running ngcc.
Waiting up to 2.5s for it to finish.
(If you are sure no ngcc process is running then you should delete the lock-file at /.../node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
Error: Timed out waiting 2.5s for another ngcc process, with id 740, to complete.
(If you are sure no ngcc process is running then you should delete the lock-file at /.../node_modules/@angular/compiler-cli/ngcc/__ngcc_lock_file__.)
    at new TimeoutError (/builds/.../node_modules/@angular/compiler-cli/ngcc/src/locking/async_locker.js:18:51)
    at AsyncLocker.<anonymous> (/builds/.../node_modules/@angular/compiler-cli/ngcc/src/locking/async_locker.js:110:31)
    at step (/builds/.../node_modules/tslib/tslib.js:143:27)
    at Object.next (/builds/.../dpf-backoffice/node_modules/tslib/tslib.js:124:57)
    at fulfilled (/builds/.../dpf-backoffice/node_modules/tslib/tslib.js:114:62)
> nx run uitkering:build:production 
An error occurred during the build:
Error: NGCC failed.

```

**Comment by localpcguy** [comment]
I am still seeing this (for us it is on parallel test runs on CI (CircleCI if it matters). It actually works ok locally via CLI depending on the computer used (more CPU = ability to run more tests in parallel). Using Angular v14.1.x, Nx v14.5.4. I verified that the ngcc command was the same as per the PRs mentioned above, and even tried some of the hacky workarounds. Effectively we need to turn off parallel for tests, as it just won't run.

I can open a new issue if that'd be better that tacking onto this older one. I can also head over the the Angular project, but I feel this probably happens more for those using Nx than for anyone else because of the ease of setting it up to run in parallel.

Related issues I've been looking at include (there was a number of other reports also, but those have the best level of information in them as far as I can tell).

- https://github.com/angular/angular/issues/38216
- https://github.com/angular/angular/issues/32431
- https://github.com/nrwl/nx/issues/6842
